File: frontend/src/types/memory.ts
export type MemoryRecordKind = "session" | "web";
export type RecentConversationView = "project" | "time";
export type RecentMemoryEventCategory =
  | "progress"
  | "decision"
  | "research"
  | "verification"
  | "blocker"
  | "follow_up";
export type MemoryRecallSessionStatus =
  "active" | "completed" | "failed" | "cancelled" | "resume_unavailable";
export type MemoryRecallTurnStatus =
  | "queued"
  | "running"
  | "completed"
  | "failed"
  | "cancelled"
  | "resume_unavailable";
export type MemoryPublicTaskStatus =
  "pending" | "running" | "cancelling" | "succeeded" | "failed" | "cancelled";

export interface RecentMemoryEvent {
  id: string;
  category: RecentMemoryEventCategory;
  title: string;
  summary: string;
  occurred_at: string;
}

export interface RecentMemorySession {
  session: { title: string; updated_at: string | null };
  project_path: string | null;
  last_activity_at: string;
  source_agent: string;
  question_count: number;
  turn_count: number;
  recent_events: RecentMemoryEvent[];
}

export interface RecentMemoryEventTarget {
  record_kind: MemoryRecordKind;
  session_id: string;
  question_id: string | null;
  turn_id: string | null;
  block_id: string | null;
}

export interface MemoryScope {
  app_id: string | null;
  source_id: string | null;
  project_path: string | null;
  session_id: string | null;
}

export interface MemoryNavigationTarget {
  record_kind: MemoryRecordKind;
  source_id: string | null;
  session_id: string;
  question_id: string | null;
  turn_id: string | null;
  part_id: string | null;
  block_id: string | null;
}

export interface MemoryRecallSearchHit {
  record_kind: MemoryRecordKind;
  source_id: string;
  session_id: string;
  session_title: string;
  project_path: string | null;
  question_id: string;
  question_index: number;
  turn_id: string | null;
  part_id: string | null;
  block_id: string;
  card_type: string;
  snippet: string;
  lexical_score: number;
  semantic_score: number;
  score: number;
  sources: string[];
}

export interface MemoryRecallSearchResult {
  query: string;
  backend: string;
  total_count: number;
  hits: MemoryRecallSearchHit[];
}

export interface MemoryRecallSessionReference {
  recordKind: MemoryRecordKind;
  sessionId: string;
  questionId: string | null;
}

export interface MemoryRecallContentReference {
  recordKind: MemoryRecordKind;
  sessionId: string;
  questionId: string;
  turnId: string | null;
  partId: string | null;
  blockId: string;
}

export interface MemoryRecallStructuredOutput {
  answer: string;
  sessionReferences: MemoryRecallSessionReference[];
  contentReferences: MemoryRecallContentReference[];
  followUpSuggestions: string[];
}

export interface MemoryRecallTurn {
  id: string;
  sessionId: string;
  sequence: number;
  conversationSessionId: string;
  conversationTurnId: string;
  status: MemoryRecallTurnStatus;
  userText: string;
  structuredOutput: MemoryRecallStructuredOutput | null;
  lastError: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface MemoryRecallSession {
  id: string;
  status: MemoryRecallSessionStatus;
  scope: MemoryScope;
  executionContextKey: string;
  agentId: string;
  model: string | null;
  turnCount: number;
  activeTurnId: string | null;
  lastError: string | null;
  createdAt: string;
  updatedAt: string;
  turns: MemoryRecallTurn[];
}

export interface MemoryContextReference {
  kind: string;
  id: string;
  source_revision: number | null;
}

export interface MemoryContextResult {
  text: string;
  revision: string;
  generated_at: string | null;
  estimated_tokens: number;
  token_budget: number;
  references: MemoryContextReference[];
  global_version: Record<string, unknown> | null;
  project_version: Record<string, unknown> | null;
  project_sources: Record<string, unknown>[];
}

export interface MemoryProjectView {
  project: Record<string, unknown>;
  version: Record<string, unknown> | null;
  sources: Record<string, unknown>[];
}

export interface MemoryRebuildResult {
  scope: MemoryScope;
  queued: boolean;
  scheduled_tasks: number;
}

export interface MemoryTaskProgress {
  current: number;
  total: number | null;
  note: string | null;
}

export interface MemoryTaskView {
  id: string;
  status: MemoryPublicTaskStatus;
  kind: string;
  progress: MemoryTaskProgress | null;
  started_at: string;
  finished_at: string | null;
  result: unknown;
  error: {
    code: string;
    message: string;
    retryable: boolean;
    details?: unknown;
  } | null;
  detail: Record<string, unknown>;
}

export type RecentMemoryStatus = "empty" | "generating" | "ready" | "update_failed";
export type RecentSnapshotPublicationKind = "generated" | "reused";
export type SourceAvailability = "available" | "partially_unavailable" | "unavailable";

export interface RecentMemoryErrorView {
  code: string;
  message: string;
  retryable: boolean;
}

export interface RecentSessionReferenceView {
  sourceId: string;
  sessionId: string;
  sessionTitle: string;
  sourceAgent: string;
  lastActivityAt: string;
  available: boolean;
  unavailableReason: string | null;
}

export interface RecentMemoryItemView {
  itemId: string;
  revisionId: string;
  category: string;
  status: string;
  title: string;
  summary: string;
  rationale: string;
  occurredAt: string;
  recommendationRank: number | null;
  sourceAvailability: SourceAvailability;
  sessionReferences: RecentSessionReferenceView[];
}

export interface RecentProjectView {
  projectKey: string;
  projectTitle: string;
  projectPath: string | null;
  summary: string;
  noMaterialChange: boolean;
  latestActivityAt: string;
  sourceSessionCount: number;
  items: RecentMemoryItemView[];
}

export interface RecentMemorySnapshotView {
  snapshotId: string;
  sequence: number;
  targetWatermark: string;
  windowStart: string;
  windowEnd: string;
  windowHours: number;
  publicationKind: RecentSnapshotPublicationKind;
  reusedFromSnapshotId: string | null;
  contentGeneratedAt: string;
  publishedAt: string;
  projects: RecentProjectView[];
}

export interface RecentMemoryStateView {
  status: RecentMemoryStatus;
  snapshot: RecentMemorySnapshotView | null;
  latestAttemptTaskId: string | null;
  latestAttemptError: RecentMemoryErrorView | null;
}

File: src-tauri/src/backend/models/memory.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum MemoryRecordKind {
    Session,
    Web,
}

impl MemoryRecordKind {
    pub(crate) fn as_str(self) -> &'static str {
        match self {
            Self::Session => "session",
            Self::Web => "web",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct MemoryRecallQuestionRef {
    pub record_kind: MemoryRecordKind,
    pub source_id: String,
    pub session_id: String,
    pub session_title: String,
    pub project_path: Option<String>,
    pub question_id: String,
    pub question_index: i64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct MemoryRecallSearchHit {
    pub record_kind: MemoryRecordKind,
    pub source_id: String,
    pub session_id: String,
    pub session_title: String,
    pub project_path: Option<String>,
    pub question_id: String,
    pub question_index: i64,
    pub turn_id: Option<String>,
    pub part_id: Option<String>,
    pub block_id: String,
    pub card_type: String,
    pub snippet: String,
    pub lexical_score: u64,
    pub semantic_score: u64,
    pub score: u64,
    pub sources: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct MemoryRecallSearchResult {
    pub query: String,
    pub backend: String,
    pub total_count: usize,
    pub hits: Vec<MemoryRecallSearchHit>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum MemoryRecallSessionStatus {
    Active,
    Completed,
    Failed,
    Cancelled,
    ResumeUnavailable,
}

impl MemoryRecallSessionStatus {
    pub(crate) fn as_str(self) -> &'static str {
        match self {
            Self::Active => "active",
            Self::Completed => "completed",
            Self::Failed => "failed",
            Self::Cancelled => "cancelled",
            Self::ResumeUnavailable => "resume_unavailable",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum MemoryRecallTurnStatus {
    Queued,
    Running,
    Completed,
    Failed,
    Cancelled,
    ResumeUnavailable,
}

impl MemoryRecallTurnStatus {
    pub(crate) fn as_str(self) -> &'static str {
        match self {
            Self::Queued => "queued",
            Self::Running => "running",
            Self::Completed => "completed",
            Self::Failed => "failed",
            Self::Cancelled => "cancelled",
            Self::ResumeUnavailable => "resume_unavailable",
        }
    }

    pub(crate) fn is_terminal(self) -> bool {
        matches!(
            self,
            Self::Completed | Self::Failed | Self::Cancelled | Self::ResumeUnavailable
        )
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct MemoryRecallSessionReference {
    pub record_kind: MemoryRecordKind,
    pub session_id: String,
    pub question_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct MemoryRecallContentReference {
    pub record_kind: MemoryRecordKind,
    pub session_id: String,
    pub question_id: String,
    pub turn_id: Option<String>,
    pub part_id: Option<String>,
    pub block_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct MemoryRecallStructuredOutput {
    pub answer: String,
    #[serde(alias = "session_references")]
    pub session_references: Vec<MemoryRecallSessionReference>,
    #[serde(alias = "content_references")]
    pub content_references: Vec<MemoryRecallContentReference>,
    #[serde(alias = "follow_up_suggestions")]
    pub follow_up_suggestions: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct MemoryRecallTurn {
    pub id: String,
    pub session_id: String,
    pub sequence: i64,
    pub conversation_session_id: String,
    pub conversation_turn_id: String,
    pub status: MemoryRecallTurnStatus,
    pub user_text: String,
    pub structured_output: Option<MemoryRecallStructuredOutput>,
    pub last_error: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct MemoryRecallSession {
    pub id: String,
    pub status: MemoryRecallSessionStatus,
    pub scope: MemoryScope,
    pub execution_context_key: String,
    pub agent_id: String,
    pub model: Option<String>,
    pub turn_count: i64,
    pub active_turn_id: Option<String>,
    pub last_error: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub turns: Vec<MemoryRecallTurn>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, Default)]
#[serde(rename_all = "camelCase")]
pub struct MemoryScope {
    pub app_id: Option<String>,
    pub source_id: Option<String>,
    pub project_path: Option<String>,
    pub session_id: Option<String>,
}

impl MemoryScope {
    pub fn fingerprint(&self) -> Result<String, String> {
        let payload = serde_json::to_vec(self).map_err(|error| error.to_string())?;
        Ok(format!("{:x}", Sha256::digest(payload)))
    }
}

/// 高密度精简证据管线冻结预算策略 (E0 冻结参数)
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct BoundedMemoryBudgetPolicy {
    /// 首包最大字符数（默认 32,000，约 8,000 tokens）
    pub initial_pack_max_chars: usize,
    /// 首包包含的精选节点上限
    pub initial_pack_node_limit: usize,
    /// 单项证据正文最大字符截断
    pub single_item_max_chars: usize,
    /// 单次工具补读响应最大字符数
    pub tool_response_max_chars: usize,
    /// 累计工具响应最大字符数
    pub tool_cumulative_max_chars: usize,
    /// 允许的工具补读调用总次数上限
    pub tool_call_limit: usize,
    /// 模型输出最大字符数
    pub agent_output_max_chars: usize,
}

impl Default for BoundedMemoryBudgetPolicy {
    fn default() -> Self {
        Self {
            initial_pack_max_chars: 32_000,
            initial_pack_node_limit: 32,
            single_item_max_chars: 2_000,
            tool_response_max_chars: 4_000,
            tool_cumulative_max_chars: 20_000,
            tool_call_limit: 10,
            agent_output_max_chars: 8_000,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::MemoryScope;

    #[test]
    fn memory_scope_fingerprint_is_stable_and_scope_sensitive() {
        let scope = MemoryScope {
            app_id: Some("codex".to_string()),
            project_path: Some("~/project".to_string()),
            ..MemoryScope::default()
        };
        let mut other = scope.clone();
        other.project_path = Some("~/other".to_string());

        assert_eq!(scope.fingerprint().unwrap(), scope.fingerprint().unwrap());
        assert_ne!(scope.fingerprint().unwrap(), other.fingerprint().unwrap());
    }
}
